### ISSUE TITLE
Broken Pipe Core Dumped in sliding-read.py

### ISSUE THREAD

**Issue by brigranger** [ndaniel/fusioncatcher#76]
Towards the end of the run for larger input files, I get a set of 6 errors (2 shown here):
```
Waiting for BLAT to finish running...
BLAT finished running.
-------------------------------------------------------------------------
Joining BLAT's output files...
Done.
Fixing PSL...
Sorting PSL...
Processing PSL...
Sorting again...
The end.
Traceback (most recent call last):
  File "/opt/fusioncatcher/v1.00/bin/sliding-read.py", line 142, in <module>
    fo.writelines(r)
IOError: [Errno 32] Broken pipe
Aborted (core dumped)
Using 32 process(es)...
Found 118 reads with poly-A/C/G/T/N tails (equal or more 16 repeat nucleotides)
Total number of input reads = 175309
Total number of reads written in the output = 175309
13.95308 % reads clipped due to low quality (less or equal than Q5) at 3' end (24461 out of 175309)!
Traceback (most recent call last):
  File "/opt/fusioncatcher/v1.00/bin/sliding-read.py", line 142, in <module>
    fo.writelines(r)
IOError: [Errno 32] Broken pipe
Aborted (core dumped)
```
Fusion catcher seems to continue on and finish and produce output, but I believe this error might be causing me downstream issues (currently preventing me from retrieving and checking the output). It seems like it may be creating hung processes? 

My initial thought would be that it's running out of memory or disk space, but I believe I'm giving FusionCatcher more than enough resources.

**Comment by ndaniel** [comment]
Hi!

Thanks for the bug report!

Please, could you let me know the exact command line for sliding-read.py ? That can be found in the log file called fusioncatcher.log.

**Comment by ndaniel** [comment]
It might be related to this:
http://newbebweb.blogspot.fi/2012/02/python-head-ioerror-errno-32-broken.html

**Comment by ndaniel** [comment]
Does the issue show up when using only one thread?

**Comment by brigranger** [comment]
Thanks for looking into it. 

I'm trying to get the sliding-read.py and blat info from the log file, but since it ends up hanging, this information is a bit hard to get at. Hopefully I'll have something by Monday.

It could be related to the issue you linked. As I mentioned, I have this error on larger files, which is also when I'm increasing the thread count as well. I thought I had success on more than 1 thread, but that might not be true. I can try only 1 thread, but it could take quite a while to process.

**Comment by ndaniel** [comment]
If it happens that you have some FASTQ files which can be shared with me and reproduce the issue it would help a lot! 

The sliding-read.py outputs its output via stdout to bowtie. I have no idea why bowtie would terminate earlier when reading from stdin (even that there are some bug reports regarding this https://github.com/BenLangmead/bowtie/issues/37 ). I really need to do some tests.

It is not clear to me if BLAT also crashes there. If it is in general is because FusionCatcher underestimates the memory required by BLAT. In general BLAT requires 1-2GB per thread so in your case at least 32GB of RAM should be free and probably it can shoot over 64 GB easily.

In general I found that running FusionCatcher with more than 16 (or 12?) threads does not bring anymore speed improvements.

**Comment by brigranger** [comment]
I've shared the fastq files with you (or will do so momentarily), as well as the stdout and stderr, which I think show the information you might need. 

Looking at the stdout and stderr, it looks like this Broken Pipe Core dump happens 6 times, and the call to sliding-read.py also happens 6 times. The first occurrence of sliding-read.py looks like this:
```
  Running: step = 402   Time: 10:28   Date: 2018-01-24 (elapsed time: 0d:12h:53m)
\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\
==> Current working directory: '/cromwell_root'
seqtk \
mergepe  \
 /cromwell_root/outDir/reads-ids_clip_star_psl_r1.fq.0 \
 /cromwell_root/outDir/reads-ids_clip_star_psl_r2.fq.0 \
|  \
sliding-read.py  \
--window 33 \
--step 15 \
-i - \
-o - \
|  \
bowtie  \
-t  \
-k 1 \
-v 2 \
-p 32 \
--tryhard  \
--chunkmbs 128 \
--suppress 2,3,4,5,6,7,8 \
 /cromwell_root/outDir/gene-gene_split_star.fa.0_bowtie_star_unique/ \
-  \
2> /dev/null \
|  \
LC_ALL=C  \
uniq  \
|  \
LC_ALL=C  \
sort  \
-u  \
--buffer-size 26G \
--parallel 32 \
-T /cromwell_root/outDir/tmp/ \
|  \
awk  \
'{n=length($0); r=substr($0,1,n-1); print r"a"; print r"b"}'  \
|  \
LC_ALL=C  \
uniq  \
|  \
LC_ALL=C  \
sort  \
-u  \
--buffer-size 26G \
--parallel 32 \
-T /cromwell_root/outDir/tmp/ \
> /cromwell_root/outDir/reads_filtered_unique_cuts_star.txt.0
--------------------------------------------------------------------------------
+-->EXECUTING...
--------------------------------------------------------------------------------
==> Execution time: 0 day(s), 0 hour(s), 0 minute(s), and 19 second(s)
```
(side note: I have been running this with 32 threads but also encountered this issue with 16 threads on this data set, but not with a smaller dataset.)

**Comment by ndaniel** [comment]
I was able to reproduce the error. This is because of a bug in `sliding-read.py`. It has been fixed here: https://github.com/ndaniel/fusioncatcher/commit/fbbdf41d704242be123555eb915d48b7702eacbd

A patch for `sliding-read.py` is here: [sliding-read.py.patch.txt](https://github.com/ndaniel/fusioncatcher/files/1688392/sliding-read.py.patch.txt)

I will apply this fix also FusionCatcher v1.00 such that re-installing it also should fix the bug.
